fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -262,13 +262,50 @@ private static void setDeprecated(DescriptorProtos.FieldDescriptorProto.Builder
      * @return the resulting records descriptor
      */
     @Nonnull
-    public static Descriptors.FileDescriptor addDefaultUnionIfMissing(Descriptors.FileDescriptor fileDescriptor) {
+    public static Descriptors.FileDescriptor addDefaultUnionIfMissing(@Nonnull Descriptors.FileDescriptor fileDescriptor) {
         if (MetaDataProtoEditor.hasUnion(fileDescriptor)) {
             return fileDescriptor;
         }
         DescriptorProtos.FileDescriptorProto fileDescriptorProto = fileDescriptor.toProto();
         DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptorProto.toBuilder();
-        fileBuilder.addMessageType(MetaDataProtoEditor.createDefaultUnion(fileDescriptor));
+        fileBuilder.addMessageType(createDefaultUnion(fileDescriptor));
+        try {
+            return Descriptors.FileDescriptor.buildFrom(fileBuilder.build(), fileDescriptor.getDependencies().toArray(new Descriptors.FileDescriptor[0]));
+        } catch (Descriptors.DescriptorValidationException e) {
+            throw new MetaDataException("Failed to add a default union", e);
+        }
+    }
+
+    /**
+     * Creates a default union descriptor for the given file descriptor if missing.
+     *
+     * <p>
+     * It will first create a default union based on the union of the stored meta-data. It will then add the new record
+     * types to the union. This will preserve the order of the fields in the union.
+     * </p>
+     *
+     * non-{@code NESTED} message types that exist in the base union to the synthetic union.
+     * @param fileDescriptor the file descriptor to create a union for
+     * @param baseUnionDescriptor the base union descriptor
+     * @return the builder for the union
+     */
+    @Nonnull
+    public static Descriptors.FileDescriptor addDefaultUnionIfMissing(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull Descriptors.Descriptor baseUnionDescriptor) {
+        if (MetaDataProtoEditor.hasUnion(fileDescriptor)) {
+            return fileDescriptor;
+        }
+        DescriptorProtos.FileDescriptorProto fileDescriptorProto = fileDescriptor.toProto();
+        DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptorProto.toBuilder();
+        DescriptorProtos.DescriptorProto.Builder unionDescriptorBuilder = createSyntheticUnion(fileDescriptor, baseUnionDescriptor);
+        for (Descriptors.Descriptor messageType : fileDescriptor.getMessageTypes()) {
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                if (!hasField(unionDescriptorBuilder, messageType)) {
+                    addFieldToUnion(unionDescriptorBuilder, messageType.getName(), messageType.getFullName());
```

Comment:
If the full name does not begin with a dot, I think we should add it. I did some work to do something like that in what I have for #508. It was something like:

```java
    @Nonnull
    private static String fullyQualifiedTypeName(@Nonnull DescriptorProtos.FileDescriptorProtoOrBuilder file,
                                                 @Nonnull String typeName) {
        if (file.hasPackage() && !file.getPackage().isEmpty()) {
            return "." + String.join(".", file.getPackage(), typeName);
        } else {
            return "." + typeName;
        }
    }
```

But I think just adding "." before `.getFullName` would be sufficient here.

Reply:
added ".". why do we need the "." though?

Reply:
The dot means, "start resolving the package name from the top rather than the bottom". Or in other words, if the protobuf compiler sees that a record has type `.a.b.M`, it will look for a message of type `M` in the `a.b` package, whereas if it sees a record type of type `a.b.M`, that will match a message of type `M` in *any* package that ends in `a.b`. In particular, if the record descriptor is in package `x.y.z`, it will match `x.y.z.a.b.M`, `x.y.a.b.M`, `x.a.b.M` and `a.b.M` (in that order, whichever is defined first). Likewise, if it is in package `x.y.a`, it will match `x.y.a.a.b.M`, `x.y.a.b.M`, `x.a.b.M`, and `a.b.M` (I think). 

I think the extra . is *probably* unnecessary in this case...except maybe the case where there is both a, like, `a.b.M` message and an `a.b.a.b.M` message and we are in package `a.b`...or something. But adding the dot makes it more explicit that we are fully qualifying the message name.

Reply:
thanks for the detailed explanation! Added the ".".